Feature/enable layers

Bluesmile82 opened the pull request:
Second Contextual-Layer PR 🏎 
(This PR contains the Feature/contextual-layers)
Also it has a big refactor extracting components and removing the map as its not indispensable
- [x] Add LayersSwitcher to toggle Layers visibility in the API
- [x] Add functionality for team Layers:
  - [x] Checkbox to add a team layer instead of a personal layer
  - [x] Display team layers with a label
  - [x] Add Custom layers
  - [x] Validation for custom layers

Commit:
Create LayersManager, getLayers from CARTO, add layers to map - WIP

Commit:
Test equals function

Commit:
Add to state if unique, style add button, remove proptypes

Bluesmile82 committed:
Create and implement Tab component

Bluesmile82 committed:
Fix problem with some layerUrls which had the x, y, z in caps

Bluesmile82 committed:
Add tabIndex to state

Bluesmile82 committed:
Reorder contextual Layers menu use Card and style

Bluesmile82 committed:
Add switch buttons and toggle the layers in the map

Bluesmile82 committed:
Update env sample

Bluesmile82 committed:
Update layers manager style

Bluesmile82 committed:
Add layer and get layers actions and update menu

Bluesmile82 committed:
Indentation

Bluesmile82 committed:
Merge branch 'develop' into feature/contextual-layers

# Conflicts:
#	src/constants/global.js

Commit:
Add toggleLayer for selected layers

Bluesmile82 committed:
Fix typos

Bluesmile82 committed:
Add a form for GFWLayers, submit and reset, allow controlled checkbox

Bluesmile82 committed:
Add and display team layers

Bluesmile82 committed:
Add labelId prop for internationalization in Checkbox and remove section div

Bluesmile82 committed:
Extract LayersForm component

Bluesmile82 committed:
Extract LayersSwitcher and remove duplicated functions

Bluesmile82 committed:
Extract TeamShowComponent

Bluesmile82 committed:
Style and indentation

Bluesmile82 committed:
Remove map and add padding to divide sections

Bluesmile82 committed:
WIP for the card footer

Bluesmile82 committed:
Revert "WIP for the card footer"

This reverts commit d6a855fb9026967b46465e7a75f27738859004fd.

Bluesmile82 committed:
Move button and teamsMode check out of the Card

Bluesmile82 committed:
Don’t try to update if you don’t have any selectedLayers

edbrett commented on src/components/ui/Checkbox.js:
```diff
@@ -1,14 +1,19 @@
     <div className="c-checkbox">
       <div className="checkbox">
-        <input type="checkbox" id={id} onChange={callback} defaultChecked={defaultChecked}/>
+        { typeof checked !== 'undefined' ? 
```
Can this not go directly into the checked={} param of the input?

Bluesmile82 replied:
We don't want checked to be false if the checked is not defined. We don't want it to have the param. So the checkbox is still uncontrolled.

edbrett commented on src/components/map/Map.js:
```diff
@@ -52,7 +56,19 @@ class Map extends React.Component {
   }
-
+    
+  updateLayers(layerUrls) {
```
I still dont like this. this.map can easily be passed outside of the component to the Layers component and be added there. This is poluting the map component and adding complexity when it is there to be reusable. All of this code should be inside the component that manages the layers.

edbrett commented on src/components/pages/settings/LayersForm.js:
```diff
@@ -0,0 +1,102 @@
+          selectedIndex={this.state.tabIndex}
+          handleTabIndexChange={this.handleTabIndexChange}
+        />
+        <form onSubmit={(e) => this.addLayers(e)}>
```
Why not use the Form component we have been using for other forms?

edbrett commented on src/components/pages/settings/LayersForm.js:
```diff
@@ -0,0 +1,102 @@
+          <Card className={"-big"}>
+            {this.state.tabIndex === 0 ? 
+              this.state.GFWLayers && this.state.GFWLayers.map((GFWlayer, i) => 
+                <div key={i}>
```
The key could go on the Checkbox components then you wouldn't need these divs

edbrett commented on src/components/pages/settings/LayersForm.js:
```diff
@@ -0,0 +1,102 @@
+              ) : null
+            }
+          </Card> 
+          <div>
```
these divs aren't needed

edbrett commented on src/components/pages/settings/LayersManager.js:
```diff
@@ -0,0 +1,23 @@
+class LayersManager extends React.Component {  
+  render() {
+    return (
+      <div className="c-layers-manager">
```
This is interesting, and also not particularly important, but I generally use .l-class as opposed to .c-class when something contains lots of other components. Layout classes handle positioning on the page etc, and .c-classes handle the finer details. Good to follow the same format through our projects so would be good to decide on this.

edbrett commented on src/components/pages/settings/LayersFormContainer.js:
```diff
@@ -0,0 +1,24 @@
+
+
+const mapStateToProps = ({ layers, teams }) => {
+    return { 
```
This again is point of preference, but as there are not a lot of props, and similar things are getting passed to LayerForm and LayerSwitcher, we could have a single Container to the LayersForm (their parent), and then pass the need props to each. This leaves the Switcher and Form as UI components and you can control the state of each more clearly. Lets discuss.

edbrett commented on src/components/pages/settings/LayersSwitcher.js:
```diff
@@ -0,0 +1,39 @@
+
+class LayersSwitcher extends React.Component {
+  componentWillMount(){
+    this.props.getLayers();
```
Should this be called on the parent (LayersManager)?

edbrett commented on src/components/pages/settings/LayersSwitcher.js:
```diff
@@ -0,0 +1,39 @@
+  render() {
+    return (
+      <div>
+        <div className="title"><FormattedMessage id={"settings.selectedLayers"} /></div>
```
could use a h3 tag here to be more expressive.

edbrett commented on src/components/pages/settings/LayersSwitcher.js:
```diff
@@ -0,0 +1,39 @@
+        <div className="title"><FormattedMessage id={"settings.selectedLayers"} /></div>
+        <div className="section">
+          {this.props.selectedLayers.map((selectedLayer, i) => (
+            <div key={i}>
```
Again can place these on the component to reduce code.

edbrett replied:
Sorry this is wrong. Ignore my comment

edbrett commented on src/components/pages/settings/Settings.js:
```diff
@@ -31,67 +32,19 @@ class Settings extends React.Component {
-          :
-          <div>
-          {(team && !editing) ?
+        <div className="l-content">
```
Nice work. Big reduction and simplification.

edbrett commented on src/components/pages/settings/TeamsForm.js:
```diff
@@ -145,8 +145,24 @@ class TeamsForm extends React.Component {
                 { team &&
-                  <div><button type="button" onClick={this.handleCancel} disabled={this.props.loading} className={`c-button -light `}><FormattedMessage id={"common.cancel"} /></button></div>}
-                <div><button type="submit" disabled={this.props.loading} className="c-button -right"><FormattedMessage id={"common.save"} /></button></div>
+                  <div>
```
Empty divs again. Not needed.

Commit:
Move get GFW and selected layers to the layers form
The fetch should be in the higher component, also it was conflicting with the loader

Commit:
No need for class

edbrett commented on src/components/pages/settings/TeamsForm.js:
```diff
@@ -145,8 +145,24 @@ class TeamsForm extends React.Component {
+                    </button>
+                  </div>
+                  }
+                <div>
```
Here too.

Commit:
Add loading state for layers and Move Loader

edbrett commented on src/components/ui/Tab.js:
```diff
@@ -0,0 +1,29 @@
+import PropTypes from 'prop-types';
+import { FormattedMessage } from 'react-intl';
+
+function Tab({ options, selectedIndex, handleTabIndexChange }) {
```
Cooooool. 😎 . Lets not forget we have this for other projects.

edbrett commented on src/styles/_react-switch-button.scss:
```diff
@@ -0,0 +1,229 @@
+/**
+ * React Switch Button control style.
```
Not a big fan of copying styles to override as often we only need to change a few classes. But if it works 👍

edbrett commented on src/styles/components/_c-layers-manager.scss:
```diff
@@ -0,0 +1,11 @@
+.c-layers-manager {
+    padding-top: 60px;
+  .section {
```
spacing.

edbrett commented on src/styles/components/_c-map.scss:
```diff
@@ -1,3 +1,9 @@
 .c-map {
   height: calc(100vh - #{$nav-height} - #{$hero-height} - #{$footer-height});
-}
+  &.-layers-container {
```
spacing

edbrett reviewed (changes requested):
I am still getting problems with using the layer manager in general on my machine. Lets review when you get to office together.

edbrett commented:
Also can we organise the settings pages better. There are way to many files in there and it is hard to know what are parents and what are child components. We can discuss when you get to office. We have been following a structure for this for other pages.

edbrett commented:
Also also, I think as some of these components are starting to become larger (and manage other components) we should be diving up the functions with comments to make is clearer what is happening.

Bluesmile82 committed:
Remove layers functionality in map - to be rescued

Bluesmile82 committed:
Styling changes

Bluesmile82 committed:
Update GFW call to filter empty title and ask for carto_id to avoid problems on select

Bluesmile82 committed:
Move LayersManayer component to the components folder

Bluesmile82 committed:
Move teams form to separate component, correct layers manager route

Bluesmile82 committed:
Custom layers WIP

Bluesmile82 committed:
Clear form after submit and add url validation

Bluesmile82 committed:
Prevent submit if there are errors

Bluesmile82 committed:
Style checkbox and layers form

Bluesmile82 committed:
Display apart teams layers and my layers

Bluesmile82 committed:
Delete layer button functionality

edbrett commented on src/styles/_base.scss:
```diff
@@ -52,3 +52,11 @@ textarea:focus,
   outline: none;
 }
+
+h3 {
```
we have a texts.scss file for things like this as base should be to normalize the css.

edbrett commented on src/styles/_base.scss:
```diff
@@ -52,3 +52,11 @@ textarea:focus,
 }
+
+h3 {
+  font-weight: bold;
```
we have a settings file for things like bold.

edbrett replied:
And always use number for font weights as font families etc may not do what you expect with bold.

Bluesmile82 committed:
Move css to the right place

Bluesmile82 committed:
Move teams show to a folder and rename teams-form to teams manager
Standardising purposes

Bluesmile82 committed:
Merge branch 'develop' into feature/enable-layers

# Conflicts:
#	src/index.scss
